keps/sig-instrumentation/1753-logs-sanitization/README.md

```diff
@@ -0,0 +1,243 @@
+
+
+### Test Plan
+
+### Graduation Criteria
+
+#### Alpha (1.19)
+- All well-known places which contain sensitive data have been tagged.
+- All external well-known types are handled by the GlobalDatapolicyMapping function.
+- It is possible to turn on logs sanitization for all Kubernetes components including API Server, Scheduler, Controller manager and kubelet using --logging-sanitization flag.
+
+#### Beta (1.20)
+- Performance overhead related to enabling dynamic logs sanitization has been reduced to 50% compared to time spent in klog library functions with this feature disabled. This should be verified using Kubernetes scalability tests.
+
+#### GA (1.21)
+- Logs sanitization is enabled by default.
+
+## Implementation History
+
+## Drawbacks
+
+## Alternatives
+
+### Static code analysis
+
+Instead of introducing optional dynamic filtering of logs at runtime we could use the same metadata to perform static code analysis. 
+
+#### Limitations of Static Analysis
+
+##### Theoretical
+
+The major theoretical limitation of static analysis is imposed by results from [decidability theory](https://www.tutorialspoint.com/automata_theory/language_decidability.htm) - given an arbitrary program written in a general-purpose programming language (one capable of simulating a Turing machine), it is impossible to examine the program algorithmically and determine if an arbitrarily chosen statement in the program will be executed when the program operates on arbitrarily chosen input data. Furthermore, there is no algorithmic way to identify those programs for which the analysis is possible; otherwise, the halting problem for Turing machines would be solvable.
+
+One major, ramification of this result concerns the distinctinction between syntactic and semantic control paths. Syntactic paths comprise all possible control paths through the flow graph. Semantic paths are those syntactic paths that can be executed. However, not all syntactic paths are executable paths. Thus, the semantic paths are a subset of the syntactic paths. In static analysis, it would be highly desirable to identify the semantic path. However, decidability results state that there is no algorithmic way to detect the semantic path through an arbitrary program written in a general purpose programming language.
+
+##### Practical 
+
+A major practical limitation of static analysis concerns array references and evaluation of pointer variables. Array subscripts and pointers are mechanisms for selecting a data item at runtime, based on previous computations performed by the program. Static analysis cannot evaluate subscripts pointers and, thus, is unable to distinguish between elements of an array or members of a list. Although it might be possible to analyze subscripts and pointers using techniques of symbolic execution, it is generally simpler and more efficient to use dynamic testing.
```

Comment:
Just to make sure I understand, you are saying that static analysis wouldn't be able to catch me logging (from your example above):

```golang
cm := ConfigMap{Data: map[string]string{"password": "123456"}}
klog.Errorf("Configmap with password: %+v", &cm)
```
https://play.golang.org/p/T-jcUceYQpb

Reply:
**tl;dr: pretty much**

I think `unable to distinguish` might be stronger than is strictly, categorically true, but it's not far from the mark.  I'd say it breaks down further into two possible points of failure: (a) failure in analysis configuration, and (b) difficulty in analysis for identification.  This snippet touches both.

(a) In the snippet above, we are storing sensitive information in `field Data map[string]string`.  From a static analysis standpoint, it's difficult to say "some key-value pairs might be sensitive."  If we said the entire field should never reach a sink, that is easier, but obviously more limiting than necessary.  But without some other prior knowledge, there's no reason to think an arbitrary `map[string]string` is containing something sensitive.  We might make some progress saying "any `map[string]string` entry with key `"password"` is sensitive."  Does that include `map[string]interface{}` or `map[interface{}]string`?  Do we need to do fuzzy-matching to include the keys `"passwd"` or `"PWORD"`?  Are we likely to produce false-positives if we match `"pass"`?

(b) Suppose we solve the issue of configuration and know `ConfigMap.Data` could contain a sensitive key-value pair.  This explicit reference in the above snippet would correspond to [ssa](https://pkg.go.dev/golang.org/x/tools/go/ssa) method `ssa.MapUpdate`.  We identify the `MapUpdate` using the key `"password"` and a non-empty value.  We could possibly mark the register as containing a secret and track it during analysis to ensure it doesn't reach a sink.

However, that requires an explicit `MapUpdate` with the key `"password"`.  It would miss any `ConfigMap` that arrived on the wire.  Even something as simple as [obfuscating the underlying type with an interface](https://play.golang.org/p/p1HVWGYEUtx) would require some level of pointer analysis to identify the argument reaching the sink.

So, while we might catch that two-line snippet, we're likely to miss much more in practice.

Reply:
Does that mean we can't use static analysis at all with klog?  All of the Infof/Warningf/Errorf functions take `args ...interface{}`.

Reply:
It would be an issue of the static type of the argument at the callsite, not in the call signature.

For instance, it would be pretty easy with `ssa` to identify the following:
```golang
func foo(data SuperSecretType) {
  klog.Infof("%v", data)
}
```
Here, the type of `data` at the `klog.Infof` call is concrete at compile time

Conversely, like the playground link in my previous comment:
```golang
func bar(data SuperSecretType) {
  leak(data)
}

func leak(data interface{}) {
  klog.Infof("%v", data)
}
```

Here, the type of `data` at the `klog.Infof` call is `interface{}` and its concrete type is not known to `ssa` at the callsite.

There could be a path forward if the function `leak` were identified as a data sink to avoid, either by manual configuration or a separate analysis step.  Pointer analysis could help here, but that requires full-program analysis and can get computationally expensive.

Reply:
We absolutely **can** use static analysis with functions that accept `args ...interface{}`. We can do this because Go's needs to perform conversions to `interface{}` and we can detect such conversions and thus determine the original type. 
Even with the second example provided by @PurelyApplied, the analyzer could infer the fact that the `SuperSecretType` was sent to the `leak` function and consequently treat the `data` argument as a source (when the callee is `bar`). I agree that with  @PurelyApplied that this requires full program analysis (most Static Analysis tools have such capabilities).

To summarize the answer provided by @PurelyApplied:
Both Static and Dynamic analysis would have a lot of problems with the original example provided by @dashpole . This is due to the fact that both Static and Dynamic analysis rely on the input supplied by users (domain experts) that identify potential sources of PII (ex. `Secret`) and sinks (ex. `klog.Infof`). 
Clearly, we can't simply tell an analyzer to consider and `map[string]string` as a source of PII - this will lead to a large number of false positives.

So how would one handle @dashpole's scenario? There are a couple of ways:
1. Discourage (through code review) placement of PII into generic types such as maps. Instead,  require use of specialized types (ex. `Secret`) to hold security sensitive data.
2. Detect data with high levels of entropy (tokens and keys) in the logging pipeline. Since this type of analysis could be handled outside of k8s it was not mentioned in this KEP.

To answer @dashpole's original question, regarding an example where Dynamic Analysis would have advantages over static Analysis, below is (a somewhat contrived) example.
At runtime, it will be easy to see that a `Secret` is being sent to `stdout`.
In principle, a static analysis tool could also catch this, but IMO it would not be the right tool for this sort of scenarios. Therefore, we should consider a combination of both analysis: static and dynamic.

```
type value interface {
 value() string
}

type Secret struct {
 v string
}

func (s *Secret) value() string {
 return s.v
}

type NonSecret struct {
 v string
}

func (s *NonSecret) value() string {
 return s.v
}

func main() {
 values := []value{
  &Secret{v: "pwd"},
  &NonSecret{v: "hello"},
 }

 for _, v := range values {
  if len(v.value()) < 4{
   fmt.Printf("%v", v)
  }
 }
}


```